x/oracle/abci.go

```diff
@@ -0,0 +1,157 @@
+		validatorClaimMap := make(map[string]types.Claim)
+
+		maxValidators := k.StakingKeeper.MaxValidators(ctx)
+		iterator := k.StakingKeeper.ValidatorsPowerStoreIterator(ctx)
+
+		defer iterator.Close()
+
+		powerReduction := k.StakingKeeper.PowerReduction(ctx)
+
+		i := 0
+		for ; iterator.Valid() && i < int(maxValidators); iterator.Next() {
+			validator := k.StakingKeeper.Validator(ctx, iterator.Value())
+
+			// Exclude not bonded validator
+			if validator.IsBonded() {
+				valAddr := validator.GetOperator()
+				validatorClaimMap[valAddr.String()] = types.NewClaim(validator.GetConsensusPower(powerReduction), 0, 0, valAddr)
+				i++
+			}
+		}
+
+		var (
+			// voteTargets defines the symbol (ticker) denoms that we require votes on
+			voteTargets      []string
+			voteTargetDenoms []string
+		)
+
+		for _, v := range params.AcceptList {
+			voteTargets = append(voteTargets, v.SymbolDenom)
+			voteTargetDenoms = append(voteTargetDenoms, v.BaseDenom)
+		}
+
+		// Clear all exchange rates
+		k.IterateExchangeRates(ctx, func(denom string, _ sdk.Dec) (stop bool) {
+			k.DeleteExchangeRate(ctx, denom)
+			return false
+		})
+
+		// Organize votes to ballot by denom
+		// NOTE: **Filter out inactive or jailed validators**
+		voteMap := k.OrganizeBallotByDenom(ctx, validatorClaimMap)
+
+		ballotDenomSlice := types.BallotMapToSlice(voteMap)
+
+		// Iterate through ballots and update exchange rates; drop if not enough votes have been achieved.
+		for _, ballotDenom := range ballotDenomSlice {
+			// Get weighted median of exchange rates
+			exchangeRate, err := Tally(ctx, ballotDenom.Ballot, params.RewardBand, validatorClaimMap)
+			if err != nil {
+				return err
+			}
+
+			// Set the exchange rate, emit ABCI event
+			k.SetExchangeRateWithEvent(ctx, ballotDenom.Denom, exchangeRate)
+		}
+
+		// update miss counting & slashing
+		voteTargetsLen := len(voteTargets)
+
+		claimSlice := types.ClaimMapToSlice(validatorClaimMap)
+		for _, claim := range claimSlice {
+			// Skip valid voters
+			if int(claim.WinCount) == voteTargetsLen {
+				continue
+			}
+
+			// Increase miss counter
+			k.SetMissCounter(ctx, claim.Recipient, k.GetMissCounter(ctx, claim.Recipient)+1)
+		}
+
+		// Distribute rewards to ballot winners
+		k.RewardBallotWinners(
+			ctx,
+			int64(params.VotePeriod),
+			int64(params.RewardDistributionWindow),
+			voteTargetDenoms,
+			claimSlice,
+		)
+
+		// Clear the ballot
+		k.ClearBallots(ctx, params.VotePeriod)
+	}
+
+	// Slash oracle providers who missed voting over the threshold and
+	// reset miss counters of all validators at the last block of slash window
+	if IsPeriodLastBlock(ctx, params.SlashWindow) {
+		k.SlashAndResetMissCounters(ctx)
+	}
+
+	return nil
+}
+
+// Tally calculates the median and returns it. It sets the set of voters to be
+// rewarded, i.e. voted within a reasonable spread from the weighted median to
+// the store. Note, the ballot is sorted by ExchangeRate.
+func Tally(
```

Comment:
Tally is very heavy function mentally, it must be moved to some place related to core logic, to keep this ABCI file cleaner.

The function requires a dedicated unit test.

Last, but not the least: the function **implicitly** modifies the incoming argument  `validatorClaimMap` thus it needs to return it among return values and document that it's not a copy.

Reply:
I have not updated anything since this change is already in your PR.